File: src/lib/books.js
export function createBookSlug(title, author) {
  return `${title}-${author}`
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "");
}

export function buildBookDocument({ title, author, level, totalPages }) {
  return {
    title: title.trim(),
    author: author.trim(),
    level: level.trim(),
    totalPages: Number(totalPages),
    slug: createBookSlug(title, author),
  };
}

export function toClientBook(entry) {
  const book = entry?.bookId && typeof entry.bookId === "object" ? entry.bookId : null;

  if (book?.title) {
    return {
      id: book._id.toString(),
      title: book.title,
      author: book.author,
      level: book.level,
      totalPages: book.totalPages,
      currentPage: entry.currentPage,
    };
  }

  if (entry?.title) {
    return {
      id: entry._id?.toString?.() || createBookSlug(entry.title, entry.author || "book"),
      title: entry.title,
      author: entry.author,
      level: entry.level,
      totalPages: entry.totalPages,
      currentPage: entry.currentPage,
    };
  }

  return null;
}

File: kid_quest/books.py
import re


def create_book_slug(title: str, author: str) -> str:
    slug = f"{title}-{author}".lower()
    slug = re.sub(r"[^a-z0-9]+", "-", slug)
    return slug.strip("-")


def build_book_document(title: str, author: str, level: str, total_pages: int) -> dict:
    return {
        "title": title.strip(),
        "author": author.strip(),
        "level": level.strip(),
        "totalPages": int(total_pages),
        "slug": create_book_slug(title, author),
    }
